### PR TITLE
Add TPC-DS smoke tests

### PR DESCRIPTION
**Pull request by hbutani**
- Issue #49 
-  Schema using `org.apache.calcite.adapter.tpcds.TpcdsSchema` from `org.apache.calcite:calcite-plus:1.28.0`
-  For queries started with `net.hydromatic.tpcds.query.Query` and then fixed generation issues
  using Spark SQL tpcds benchmark queries.

- Applied the [windowing pull request](https://github.com/weijietong/substrait-java/tree/issue_48)
- Added `SUBSTRING` to `io.substrait.isthmus.expression.FunctionMappings`
- See `resources/tpcds/README.md` for more details; [tpcds_investigate branch for query generation and testing process](https://github.com/hbutani/substrait-java/tree/tpcds_investigate)

**Queries passing:**
```
[1, 3, 4, 6, 7, 8, 10, 11, 13, 14, 15, 16, 18, 19, 21, 22, 23, 25, 26, 28, 29, 30, 31, 32, 33, 34, 35, 37, 38, 41,
 42, 43, 45, 46, 48, 50, 52, 54, 55, 56, 58, 59, 60, 61, 62, 64, 65, 68, 69, 71, 73, 74, 79, 81, 82, 83, 85, 87, 88,
 90, 92, 94, 95, 96, 99]
```

**Queries failing:**
  - No support for reference to Aggregate expr in windowing expressions for example
    in `q12` `sum(sum(ws_ext_sales_price)) over (partition by i_class)`
    - `[12, 20, 47, 51, 53, 57, 63, 89, 98]`
  - No support for `is null` or `is not null` check
    - `[40, 44, 49, 75, 76, 77, 78, 80, 93, 97]`
  - Error on `rank()` call
    - `[36, 67, 86]`
  - Missing functions: round, STDDEV_SAMP, UPPER, LIKE
    - `[2, 17, 24, 39, 91]`
  - Concat: vargs behavior, implicit casting
    - `[5, 54, 84]`
  - Unsupported subquery nested in Project relational operator
    - `[9]`
  - Support for GROUPING
    - `[27]`
  - Cannot apply '+' to arguments of type `<DATE> + <INT>`
    - `[72]`

### CODE REVIEW

File: isthmus/src/test/java/io/substrait/isthmus/TpcdsQueryNoValidation.java [hunk 1]
```diff
@@ -20,4 +29,25 @@ public void tpcds(int query) throws Exception {
     var plan = s.execute(asString(String.format("tpcds/queries/%02d.sql", query)), creates);
     System.out.println(JsonFormat.printer().print(plan));
   }
+
```
Comment (guykhazma) [inline isthmus/src/test/java/io/substrait/isthmus/TpcdsQueryNoValidation.java:32]: Can we remove the original `tpcds` testing function as it is no longer needed?
In addition, when removing this function the partial [`schema.sql`](https://github.com/substrait-io/substrait-java/blob/main/isthmus/src/test/resources/tpcds/schema.sql) definition can be removed as well.

File: isthmus/src/test/java/io/substrait/isthmus/TpcdsQueryNoValidation.java [hunk 2]
```diff
@@ -2,11 +2,20 @@
...
 import java.util.Arrays;
+import org.apache.calcite.adapter.tpcds.TpcdsSchema;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.ValueSource;
 
 public class TpcdsQueryNoValidation extends PlanTestBase {
 
+  private void testQuery(int i) throws Exception {
+    SqlToSubstrait s = new SqlToSubstrait();
+    TpcdsSchema schema = new TpcdsSchema(1.0);
```
Comment (guykhazma) [inline isthmus/src/test/java/io/substrait/isthmus/TpcdsQueryNoValidation.java:13]: This is nice! maybe we should do the same for the TPCH schema instead of the [`schema.sql`](https://github.com/substrait-io/substrait-java/blob/main/isthmus/src/test/resources/tpch/schema.sql) we have currently.

**Review by guykhazma — COMMENTED** [review]
Looks good! some minor comments.

File: isthmus/build.gradle.kts [hunk 3]
```diff
@@ -29,6 +29,7 @@ dependencies {
   implementation("com.google.code.findbugs:jsr305:3.0.2")
   implementation("com.github.ben-manes.caffeine:caffeine:3.0.4")
   implementation("org.immutables:value-annotations:2.8.8")
+  testImplementation("org.apache.calcite:calcite-plus:1.28.0")
```
Comment (JamesRTaylor) [inline isthmus/build.gradle.kts:32]: Do we need this new dependency?

File: isthmus/src/test/resources/tpcds/README.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
+
+## Setup of Schema and Queries
+- Schema using `org.apache.calcite.adapter.tpcds.TpcdsSchema` from `org.apache.calcite:calcite-plus:1.28.0`
+- For queries started with `net.hydromatic.tpcds.query.Query` and then fixed generation issues
+  using Spark SQL tpcds benchmark queries.
+
+## Status of testing on 6/26/2022
```
Comment (JamesRTaylor) [inline isthmus/src/test/resources/tpcds/README.md:7]: I'd recommend filing github issues (one per bullet point below), then removing this section as it'll quickly get out of date (and be a pain to keep in sync).

File: isthmus/src/test/java/io/substrait/isthmus/TpcdsQueryNoValidation.java (hunk 2 above)
Comment (JamesRTaylor) [reply isthmus/src/test/java/io/substrait/isthmus/TpcdsQueryNoValidation.java:13]: 👍

File: isthmus/src/main/java/io/substrait/isthmus/expression/FunctionMappings.java [hunk 5]
```diff
@@ -35,7 +35,9 @@ public class FunctionMappings {
                 s(SqlStdOperatorTable.MINUS_DATE, "subtract"),
                 s(SqlStdOperatorTable.DATETIME_PLUS, "add"),
                 s(SqlStdOperatorTable.EXTRACT, "extract"),
-                s(SqlStdOperatorTable.LIKE))
+                s(SqlStdOperatorTable.LIKE),
+                s(SqlStdOperatorTable.SUBSTRING, "substring"),
+                s(SqlStdOperatorTable.CONCAT, "concat"))
```
Comment (JamesRTaylor) [inline isthmus/src/main/java/io/substrait/isthmus/expression/FunctionMappings.java:40]: Move this to a separate PR

File: isthmus/build.gradle.kts (hunk 3 above)
Comment (hbutani) [reply isthmus/build.gradle.kts:32]: Yes I agree about your point on dependencies.
But
- this is only a test dependency.
- It has scale factor Stats.
- Down the line if we want to develop the Query Generator the SF and stats maybe useful...

But
- Above are not very compelling, I know.
- Immediate hurdle is I have to figure out how to export the DDL script from the Tables provided by TpcdsSchema.
   Let me know if you have any pointers on this.

@guykhazma you ok with removing TpcdsSchema. You were suggesting going the opposite way of doing the same for Tpch.

File: isthmus/src/test/resources/tpcds/README.md (hunk 4 above)
Comment (hbutani) [reply isthmus/src/test/resources/tpcds/README.md:7]: Added

File: isthmus/build.gradle.kts (hunk 3 above)
Comment (jacques-n) [reply isthmus/build.gradle.kts:32]: If the class is used by any code, it will be included in the Graal image, whether it is referred to from Isthmus or just from code we reference (e.g. internal Calcite classes). I'd guess that Schema falls into that category.

File: isthmus/src/test/java/io/substrait/isthmus/TpcdsQueryNoValidation.java (hunk 1 above)
Comment (hbutani) [reply isthmus/src/test/java/io/substrait/isthmus/TpcdsQueryNoValidation.java:41]: Moved the test to a new file: `KeyConstraintsTest`

File: isthmus/build.gradle.kts (hunk 3 above)
Comment (hbutani) [reply isthmus/build.gradle.kts:32]: Based on this have left the test dependency in the updated patch.

**Review by JamesRTaylor — APPROVED** [review]
Nice work!

File: isthmus/src/main/java/io/substrait/isthmus/expression/FunctionMappings.java (hunk 5 above)
Comment (JamesRTaylor) [inline isthmus/src/main/java/io/substrait/isthmus/expression/FunctionMappings.java:40]: Rebase on main and do a force push instead of pulling this in here.

Comment (hbutani) [reply isthmus/src/main/java/io/substrait/isthmus/expression/FunctionMappings.java]: Ok, done

Comment (JamesRTaylor) [reply isthmus/src/main/java/io/substrait/isthmus/expression/FunctionMappings.java]: Still doesn’t look right. There should be a single commit. If you rebase your local branch on an updated main branch, squash the commits into a single commit, and force push back here, it should fix it up.

**Review by JamesRTaylor — APPROVED** [review]
LGTM

**Review by jacques-n — APPROVED** [review]
This is awesome @hbutani . Thanks for all the legwork!
